### PR TITLE
Remove `remap_env_constness` in queries

### PR DESCRIPTION
**Pull request by fee1-dead**
This removes some of the complexities with const traits. #88119 used to be caused by this but was fixed by `param_env = param_env.without_const()`.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/query/type_op/prove_predicate.rs [hunk 1]
```diff
@@ -32,14 +32,8 @@ impl<'tcx> super::QueryTypeOp<'tcx> for ProvePredicate<'tcx> {
...
-        mut canonicalized: Canonical<'tcx, ParamEnvAnd<'tcx, Self>>,
+        canonicalized: Canonical<'tcx, ParamEnvAnd<'tcx, Self>>,
     ) -> Fallible<CanonicalQueryResponse<'tcx, ()>> {
-        match canonicalized.value.value.predicate.kind().skip_binder() {
-            ty::PredicateKind::Clause(ty::Clause::Trait(pred)) => {
-                canonicalized.value.param_env.remap_constness_with(pred.constness);
-            }
-            _ => canonicalized.value.param_env = canonicalized.value.param_env.without_const(),
-        }
         tcx.type_op_prove_predicate(canonicalized)
```
Comment (workingjubilee) [inline compiler/rustc_trait_selection/src/traits/query/type_op/prove_predicate.rs:37]: Hmm. Is it possible this was altering canonicalization in a way that didn't _just_ have to do with the param environment's constness?

Comment (oli-obk) [reply compiler/rustc_trait_selection/src/traits/query/type_op/prove_predicate.rs:37]: No, that's not possible.

**Pull request merged**